src/hypofuzz/dashboard/dashboard.py

```diff
@@ -454,31 +479,70 @@ async def handle_event(receive_channel: MemoryReceiveChannel[ListenerEventT]) ->
+                if event.value.property not in TESTS:
+                    continue
+                TESTS[event.value.property].failure = event.value
+            elif event.key is DatabaseEventKey.ROLLING_OBSERVATION:
+                if event.value.property not in TESTS:
+                    continue
+                TESTS[event.value.property].rolling_observations.append(event.value)
+            elif event.key is DatabaseEventKey.CORPUS_OBSERVATION:
+                if event.value.property not in TESTS:
+                    continue
+                TESTS[event.value.property].corpus_observations.append(event.value)
+
+            await broadcast_event(event.type, event.key, event.value)
+
+        # we're handling deletion events in a customish way, since event.value is
+        # always None for databases which don't support value deletion. The value
+        # we send to the websocket .on_event method is computed here and is specific
+        # to each event type.
         if event.type == "delete":
-            # don't send deletion events to the dashboard, for now
-            continue
-
-        assert event.type == "save"
-        assert event.value is not None
-
-        if event.key is DatabaseEventKey.REPORT:
-            if event.value.nodeid not in TESTS:
-                continue
-            TESTS[event.value.nodeid].add_report(event.value)
-        elif event.key is DatabaseEventKey.FAILURE_OBSERVATION:
-            if event.value.property not in TESTS:
-                continue
-            TESTS[event.value.property].failure = event.value
-        elif event.key is DatabaseEventKey.ROLLING_OBSERVATION:
-            if event.value.property not in TESTS:
-                continue
-            TESTS[event.value.property].rolling_observations.append(event.value)
-        elif event.key is DatabaseEventKey.CORPUS_OBSERVATION:
-            if event.value.property not in TESTS:
-                continue
-            TESTS[event.value.property].corpus_observations.append(event.value)
-
-        await broadcast_event(event.type, event.key, event.value)
+            if event.key is DatabaseEventKey.FAILURE_OBSERVATION:
+                # we know a failure was just deleted, but not from which test (unless
+                # the db supports value deletion). Re-scan all failing tests.
+                for test in TESTS.values():
+                    if test.failure is None:
+                        continue
+                    failure_observations = get_failure_observations(
+                        test.database_key_bytes
+                    )
+                    if not failure_observations.values():
+                        previous_failure = test.failure
+                        test.failure = None
+                        await broadcast_event(
+                            "delete",
+                            DatabaseEventKey.FAILURE_OBSERVATION,
+                            previous_failure,
+                        )
```

Comment:
I think we can do a lot better than a full scan here!  The `event.database_key` includes the db key of the failing test, and we should be find the Test instance with a dict lookup at that point.  Also if the DB _does_ support value deletion, we're not yet using that here.

Reply:
damn, I even had a moment of "it sure would be nice to know the database key here, but it's just not possible in general", except it totally is lol.

and yeah there's a bunch of places we can do better with value deletion. I don't think I'm going to attempt them yet, because I don't have a good way to test it, so the code is likely to rot. I'll open an issue to track it